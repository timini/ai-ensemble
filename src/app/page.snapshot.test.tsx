import { render } from '@testing-library/react';
import Home from './page';
import { vi } from 'vitest';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { httpBatchLink } from '@trpc/client';
import { api } from '~/trpc/react';

// Create a test query client
const createTestQueryClient = () => new QueryClient({
  defaultOptions: {
    queries: {
      retry: false,
    },
    mutations: {
      retry: false,
    },
  },
});

// Create test tRPC client
const createTestTRPCClient = (queryClient: QueryClient) => {
  return api.createClient({
    links: [
      httpBatchLink({
        url: 'http://localhost:3000/api/trpc',
        fetch: vi.fn().mockResolvedValue({
          ok: true,
          json: () => Promise.resolve({ result: { data: { isValid: true } } })
        } as Response),
      }),
    ],
  });
};

// Test wrapper component
function TestWrapper({ children }: { children: React.ReactNode }) {
  const queryClient = createTestQueryClient();
  const trpcClient = createTestTRPCClient(queryClient);

  return (
    <api.Provider client={trpcClient} queryClient={queryClient}>
      <QueryClientProvider client={queryClient}>
        {children}
      </QueryClientProvider>
    </api.Provider>
  );
}

// Mock localStorage for the test
const localStorageMock = {
  getItem: vi.fn((key: string) => {
    switch (key) {
      case 'ai-ensemble-selected-models':
        return JSON.stringify([
          { id: 'openai-gpt-4o-123', name: 'OpenAI GPT-4o', provider: 'openai', model: 'gpt-4o' },
          { id: 'google-gemini-2.5-pro-456', name: 'Google Gemini 2.5 Pro', provider: 'google', model: 'gemini-2.5-pro' },
          { id: 'anthropic-claude-3.5-sonnet-789', name: 'Anthropic Claude 3.5 Sonnet', provider: 'anthropic', model: 'claude-3-5-sonnet-20241022' }
        ]);
      case 'ai-ensemble-prompt':
        return 'Explain the concept of artificial intelligence and its impact on society in detail.';
      case 'ai-ensemble-summarizer':
        return 'openai-gpt-4o-123';
      case 'ai-ensemble-provider-keys':
        return JSON.stringify({
          openai: 'sk-test-key-123',
          google: 'test-google-key-456',
          anthropic: 'test-anthropic-key-789',
          grok: ''
        });
      case 'ai-ensemble-provider-status':
        return JSON.stringify({
          openai: { isValid: true, isValidating: false },
          google: { isValid: true, isValidating: false },
          anthropic: { isValid: true, isValidating: false },
          grok: { isValid: false, isValidating: false }
        });
      case 'ai-ensemble-available-models':
        return JSON.stringify({
          openai: ['gpt-4o', 'gpt-4', 'gpt-3.5-turbo'],
          google: ['gemini-2.5-pro', 'gemini-2.0-flash-exp', 'gemini-1.5-pro'],
          anthropic: ['claude-3-5-sonnet-20241022', 'claude-3-5-haiku-20241022', 'claude-3-opus-20240229'],
          grok: ['grok-beta']
        });
      default:
        return null;
    }
  }),
  setItem: vi.fn(),
  removeItem: vi.fn(),
  clear: vi.fn()
};

// Mock fetch for API calls
global.fetch = vi.fn();

describe('Home Page Snapshot', () => {
  beforeEach(() => {
    // Mock localStorage
    Object.defineProperty(window, 'localStorage', {
      value: localStorageMock,
      writable: true
    });

    // Mock fetch for validation endpoints
    (global.fetch as ReturnType<typeof vi.fn>).mockImplementation((url: string) => {
      if (url.includes('/api/trpc/validation.validateApiKey')) {
        return Promise.resolve({
          ok: true,
          json: () => Promise.resolve({ result: { data: { isValid: true } } })
        } as Response);
      }
      if (url.includes('/api/trpc/validation.getModels')) {
        return Promise.resolve({
          ok: true,
          json: () => Promise.resolve({ 
            result: { 
              data: ['gpt-4o', 'gpt-4', 'gpt-3.5-turbo'] 
            } 
          })
        } as Response);
      }
      return Promise.reject(new Error('Unknown URL'));
    });

    // Suppress console warnings during tests
    vi.spyOn(console, 'warn').mockImplementation(() => {});
    vi.spyOn(console, 'error').mockImplementation(() => {});
  });

  afterEach(() => {
    vi.restoreAllMocks();
    vi.clearAllMocks();
  });

  it('should match snapshot of the complete home page with data', async () => {
    const { container } = render(
      <TestWrapper>
        <Home />
      </TestWrapper>
    );
    
    // Wait a bit for the component to initialize with localStorage data
    await new Promise(resolve => setTimeout(resolve, 100));
    
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render the main page component', () => {
    const { container } = render(
      <TestWrapper>
        <Home />
      </TestWrapper>
    );
    expect(container.firstChild).toBeDefined();
  });

  it('should render without crashing with mocked localStorage data', () => {
    expect(() => render(
      <TestWrapper>
        <Home />
      </TestWrapper>
    )).not.toThrow();
  });
});
