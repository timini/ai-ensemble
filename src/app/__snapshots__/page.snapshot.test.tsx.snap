// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Home Page Snapshot > should match snapshot of the complete home page with data 1`] = `
<main
  class="flex min-h-screen flex-col items-center bg-gray-900 text-white p-4 md:p-8"
>
  <div
    class="w-full max-w-6xl"
  >
    <h1
      class="text-4xl md:text-5xl font-extrabold tracking-tight text-center mb-4"
    >
      AI 
      <span
        class="text-[hsl(280,100%,70%)]"
      >
        Ensemble
      </span>
    </h1>
    <p
      class="text-center text-gray-400 mb-8"
    >
      Compare. Consensus. Confidence.
    </p>
    <div
      class="space-y-6"
    >
      <div
        class="bg-gray-800 rounded-lg p-6"
      >
        <h2
          class="text-xl font-bold mb-4"
        >
          Configuration
        </h2>
        <div
          class="flex flex-wrap gap-4"
        >
          <div
            class="relative"
          >
            <button
              class="flex items-center gap-2 bg-gray-700 hover:bg-gray-600 text-white px-4 py-2 rounded-lg transition-colors"
            >
              <span>
                ðŸ”§
              </span>
              <span>
                Configure Providers (
                0
                /4)
              </span>
              <span
                class="transform transition-transform "
              >
                â–¼
              </span>
            </button>
          </div>
          <div
            class="relative"
          >
            <button
              class="flex items-center gap-2 bg-blue-600 hover:bg-blue-700 text-white px-4 py-2 rounded-lg transition-colors"
            >
              <span>
                ðŸŽ¯
              </span>
              <span>
                Select Models (
                0
                /8)
              </span>
              <span
                class="transform transition-transform "
              >
                â–¼
              </span>
            </button>
          </div>
        </div>
      </div>
      <div
        class="bg-gray-800 border-2 border-dashed border-gray-600 rounded-lg p-8 text-center"
      >
        <div
          class="text-gray-400 mb-2"
        >
          ðŸŽ¯
        </div>
        <h3
          class="text-lg font-medium text-gray-300 mb-2"
        >
          No Models Selected
        </h3>
        <p
          class="text-gray-400"
        >
          Configure providers and select 2-8 AI models to start comparing responses.
        </p>
      </div>
      <form
        class="space-y-4"
      >
        <textarea
          class="bg-gray-800 p-3 rounded-lg text-lg focus:outline-none focus:ring-2 focus:ring-[hsl(280,100%,70%)] w-full"
          placeholder="Enter your prompt here..."
          rows="5"
        />
        <div
          class="flex flex-col items-end gap-2"
        >
          <div
            class="text-sm text-amber-400 text-right"
          >
            Select at least 2 AI models to start
          </div>
          <button
            class="bg-[hsl(280,100%,70%)] text-white font-bold py-2 px-6 rounded-lg hover:bg-[hsl(280,100%,60%)] disabled:bg-gray-600 disabled:cursor-not-allowed"
            disabled=""
            type="submit"
          >
            Compare 0 Models
          </button>
        </div>
      </form>
    </div>
  </div>
</main>
`;

exports[`Home Page Snapshot > should match snapshot of the home page structure 1`] = `
<div
  data-testid="improved-ensemble-interface"
>
  <h1>
    AI Ensemble - Compare Multiple AI Models
  </h1>
  <div
    data-testid="configuration-section"
  >
    <h2>
      Configuration
    </h2>
    <div
      data-testid="provider-config"
    >
      Provider Configuration
    </div>
    <div
      data-testid="model-selection"
    >
      Model Selection
    </div>
  </div>
  <div
    data-testid="selected-models"
  >
    Selected Models Display
  </div>
  <div
    data-testid="query-form"
  >
    <div
      data-testid="prompt-input"
    >
      Prompt Input
    </div>
    <div
      data-testid="summarizer-selection"
    >
      Summarizer Selection
    </div>
    <button>
      Compare Models
    </button>
  </div>
  <div
    data-testid="results-section"
  >
    <div
      data-testid="consensus-response"
    >
      Consensus Response
    </div>
    <div
      data-testid="agreement-analysis"
    >
      Agreement Analysis
    </div>
    <div
      data-testid="individual-responses"
    >
      Individual Responses
    </div>
  </div>
</div>
`;
